### ISSUE TITLE
load_state_dict doesn't catch 'StopIteration' exception

### ISSUE THREAD

**Issue by zzzacwork** [lhotse-speech/lhotse#753]
Hi everyone,

I am using icefall to train a RNNT model and after I start from a batch and load the sampler sate, the dynamicbucketsampler raise an StopIteration which resulted the program to abort. I think that "exception" should have been caught somewhere up the stack trace and processed. 

```
-- Process 0 terminated with the following error:
Traceback (most recent call last):
  File "/home/azureuser/anaconda3/envs/icefall-torch1.10cu102/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
    fn(i, *args)
  File "/home/azureuser/codebase/icefall-birch/egs/gigaspeech/ASR/pruned_transducer_stateless2/train.py", line 869, in run
    train_dl = gigaspeech.train_dataloaders(
  File "/home/azureuser/codebase/icefall-birch/egs/gigaspeech/ASR/pruned_transducer_stateless2/asr_datamodule.py", line 320, in train_dataloaders
    train_sampler.load_state_dict(sampler_state_dict)
  File "/home/azureuser/codebase/lhotse-birch/lhotse/dataset/sampling/dynamic_bucketing.py", line 174, in load_state_dict
    self._fast_forward()
  File "/home/azureuser/codebase/lhotse-birch/lhotse/dataset/sampling/dynamic_bucketing.py", line 190, in _fast_forward
    next(self)
  File "/home/azureuser/codebase/lhotse-birch/lhotse/dataset/sampling/base.py", line 261, in __next__
    batch = self._next_batch()
  File "/home/azureuser/codebase/lhotse-birch/lhotse/dataset/sampling/dynamic_bucketing.py", line 233, in _next_batch
    batch = next(self.cuts_iter)
StopIteration
```

**Comment by pzelasko** [comment]
Is this the full stack trace, or were there more parts? Is it possible the actual error was somewhere else? 

Also, at which step was the sampler state dict saved? Was it a continued training on the same data?

**Comment by zzzacwork** [comment]
Sorry for the lack of context... 

It might be more appropriate for an issue within icefall, I am fine-tuning a pretrained model under the gigaspeech recipe, and I restarted the training after it failed after around 18000 batches. and then I noticed this error. The state dict was saved from [the train.py script](https://github.com/k2-fsa/icefall/blob/dc89b61b808a22b64e7e38cb09771ea7a6bb64d2/egs/gigaspeech/ASR/pruned_transducer_stateless2/train.py#L700). I think the saving and loading of the model seems working and the behavior of the next function to raise an exception is also. normal, it might just need to be caught and handled in some way.  More stack info was pasted below,

```
2022-06-21 20:49:26,257 INFO [asr_datamodule.py:319] (2/4) Loading sampler state dict                                                       [881/1541]
2022-06-21 20:49:26,502 INFO [asr_datamodule.py:316] (3/4) About to create train dataloader
2022-06-21 20:49:26,502 INFO [asr_datamodule.py:319] (3/4) Loading sampler state dict                                                                 


Traceback (most recent call last):                                         
  File "./pruned_transducer_stateless2/train.py", line 996, in <module>                                                                               
    main()                                                                 
  File "./pruned_transducer_stateless2/train.py", line 987, in main                                                                                   
    mp.spawn(run, args=(world_size, args), nprocs=world_size, join=True)                                                                              
  File "/home/azureuser/anaconda3/envs/icefall-torch1.10cu102/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 230, in spawn         
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')                                                                      
  File "/home/azureuser/anaconda3/envs/icefall-torch1.10cu102/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 188, in start_processe
s                                    
    while not context.join():                                              
  File "/home/azureuser/anaconda3/envs/icefall-torch1.10cu102/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 150, in join          
    raise ProcessRaisedException(msg, error_index, failed_process.pid)                                                                                
torch.multiprocessing.spawn.ProcessRaisedException:                                                                                                   

-- Process 1 terminated with the following error:                                                                                                     
Traceback (most recent call last):                                         
  File "/home/azureuser/anaconda3/envs/icefall-torch1.10cu102/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap          
    fn(i, *args)                                                           
  File "/home/azureuser/codebase/icefall-birch/egs/gigaspeech/ASR/pruned_transducer_stateless2/train.py", line 869, in run                            
    train_dl = gigaspeech.train_dataloaders(                                                                                                          
  File "/home/azureuser/codebase/icefall-birch/egs/gigaspeech/ASR/pruned_transducer_stateless2/asr_datamodule.py", line 320, in train_dataloaders     
    train_sampler.load_state_dict(sampler_state_dict)                                                                                                 
  File "/home/azureuser/codebase/lhotse-birch/lhotse/dataset/sampling/dynamic_bucketing.py", line 174, in load_state_dict
    self._fast_forward()                                                   
  File "/home/azureuser/codebase/lhotse-birch/lhotse/dataset/sampling/dynamic_bucketing.py", line 190, in _fast_forward
    next(self)                                                             
  File "/home/azureuser/codebase/lhotse-birch/lhotse/dataset/sampling/base.py", line 261, in __next__
    batch = self._next_batch()                                             
  File "/home/azureuser/codebase/lhotse-birch/lhotse/dataset/sampling/dynamic_bucketing.py", line 233, in _next_batch
    batch = next(self.cuts_iter)                                           
StopIteration    
```

**Comment by zzzacwork** [comment]
This error still persists when I try to restart the training in icefall using --start-batch parameter, I looked into the code and the following part from `sampling/dynamic_bucketing.py` a bit confusing. if I understand correctly it tries to iterate to `total_batches` from that epoch, instead of the batch that I try to restart from. Then it causes the sampler to run to the end raise an exception. Should we change it to iterate to the current batch index instead?


```
    def _fast_forward(self):
        current_epoch = self.diagnostics.current_epoch
        num_batches_to_iter = self.diagnostics.current_epoch_stats.total_batches

        # Set the right epoch
        self.set_epoch(current_epoch)
        # Reset diagnostics for this epoch as we're about to re-iterate
        self.diagnostics.stats_per_epoch[current_epoch] = EpochDiagnostics(
            epoch=current_epoch
        )

        self._just_restored_state = False
        iter(self)
        for _ in range(num_batches_to_iter):
            next(self)
        self._just_restored_state = True
```

**Comment by zzzacwork** [comment]
Thank you for the prompt reply.
below is the info from `get_report`
```
| ep   0 | cuts kept 28408701/28408701 (100.00%) | cuts discarded 0 | batches kept 2826061/2826061 (100.00%)| batches discarded 0 |
|  total  | cuts kept 28408701/28408701 (100.00%) | cuts discarded 0 | batches kept 2826061/2826061 (100.00%)| batches discarded 0 |
```

The checkpoint I tried to load is from epoch 0, batch 21000(`checkpoint-210000.pt`)

**Comment by pzelasko** [comment]
Can you show the contents of sampler's state dict in `checkpoint-210000.pt`?

**Comment by zzzacwork** [comment]
```
{'epoch': 0, 'drop_last': True, 'world_size': 4, 'rank': 0, 'seed': 0, 'shuffle': True, 'diagnostics': {'current_epoch': 0, 'stats_per_epoch': {0: {'epoch': 0, 'kept_cuts': 28408701, 'discarded_cuts': 0, 'kept_batches': 2826061, 'discarded_batches': 0}}}, 'max_duration': 100, 'max_cuts': None, 'consistent_ids': True, 'buffer_size': 10000, 'num_cuts_for_bins_estimate': 10000, 'shuffle_buffer_size': 20000, 'strict': True}
```
it seems the kept_batches was saved wrongly?

**Comment by pzelasko** [comment]
Yeah, that seems to be the case. Can you try to manually overwrite it and see if it works as expected then?

**Comment by zzzacwork** [comment]
I tried to manually set `num_batches_to_iter = 210000` and it seems to work. 

`kept_batches` exceeds the total number of batches in the dataset I think.

**Comment by zzzacwork** [comment]
I checked the next checkpoint for instance, checkpoint-220000 and its kept_batches is 
2866061, 40000 more than checkpoint-210000.pt, which makes sense since I am training it on a 4GPU machine. 

I rerun it from the beginning, and add a line to print the `get_report` when saving checkpoint. 

```
| ep   0 | cuts kept 42601/42601 (100.00%) | cuts discarded 0 | batches kept 4008/4008 (100.00%)| batches discarded 0 | 
| ep   0 | cuts kept 77808/77808 (100.00%) | cuts discarded 0 | batches kept 8008/8008 (100.00%)| batches discarded 0 |
| ep   0 | cuts kept 113140/113140 (100.00%) | cuts discarded 0 | batches kept 12008/12008 (100.00%)| batches discarded 0 | 
```
there is a difference of 8 batches, is that normal?

**Comment by pzelasko** [comment]
Yes, a difference of prefetch factor * number of GPUs is expected because the dataloader fetches some batches ahead of the time when they are actually used

**Comment by zzzacwork** [comment]
I then stop and restart from checkpoint-3000, and find the discrepancy at the next checkpoint,(checkpoint-4000)
```
| ep   0 | cuts kept 569182/569182 (100.00%) | cuts discarded 0 | batches kept 64036/64036 (100.00%)| batches discarded 0 |
|  total  | cuts kept 569182/569182 (100.00%) | cuts discarded 0 | batches kept 64036/64036 (100.00%)| batches discarded 0 |
2022-07-07 23:19:03,113 INFO [checkpoint.py:75] (0/4) Saving checkpoint to birch/exp_0707/checkpoint-4000.pt
```
which should be 16008

**Comment by zzzacwork** [comment]
I narrowed down the problem by checking the diagnostics right after `load_state`, 
![image](https://user-images.githubusercontent.com/102050116/177896724-346c0b4d-a1df-48cd-a775-0782ef5d59df.png)

The first 8008 is printed before `_fast_forward`, the 32032 is printed after `load_state`. it seems the number is multiplied by the world size, but I cannot find out where.

**Comment by pzelasko** [comment]
Oh I understand what's going wrong now. Because of world_size > 1, the reported number of batches kept is `world_size` times larger than the actual number of returned batches. Then we try to iterate and `next()` is returning every `world_size`-th batch so in case `world_size=4` it over-iterates 4 times. I will fix it tomorrow.

**Comment by pzelasko** [comment]
@zzzacwork can you try our this PR and let me know if it's better? Unfortunately, you will need to generate a fresh checkpoint, because I fixed the x4 multiplier thing when counting the number of batches kept.

https://github.com/lhotse-speech/lhotse/pull/773

**Comment by zzzacwork** [comment]
Hi @pzelasko , Thanks a lot for the prompt fix! 
I did a test on that PR and found the initial batches kept parameter was changed to be equal to the batch index. However, the actual batches sampled should times world_size I think. I guess the bug is due to the fact that  the `fast_forward` probably should have a different way calling the `keep` function. 
![image](https://user-images.githubusercontent.com/102050116/178011307-9c6a215a-bd87-4a8a-8cf6-1254dc0cef8b.png)
